src/idpi/grib_decoder.py

```diff
@@ -91,227 +94,297 @@ def _extract_pv(pv):
     }
 
 
-def _update_origin(metadata, ref_param):
-    x0_key = "longitudeOfFirstGridPointInDegrees"
-    y0_key = "latitudeOfFirstGridPointInDegrees"
-    ref_geo = metadata[ref_param]["geography"]
-    x0 = ref_geo[x0_key] % 360
-    y0 = ref_geo[y0_key]
-    for m in metadata.values():
-        geo = m["geography"]
-        dx = geo["iDirectionIncrementInDegrees"]
-        dy = geo["jDirectionIncrementInDegrees"]
-        m["origin"] |= {
-            "x": np.round((geo[x0_key] % 360 - x0) / dx, 1),
-            "y": np.round((geo[y0_key] - y0) / dy, 1),
-        }
+@dc.dataclass
+class Grid:
+    lon: xr.DataArray
+    lat: xr.DataArray
+    longitudeOfFirstGridPointInDegrees: float
+    latitudeOfFirstGridPointInDegrees: float
+
+
+def _check_string_arg(obj):
+    return bool(obj) and all(isinstance(elem, str) for elem in obj)
+
+
+class GribReader:
+    def __init__(
+        self,
+        datafiles: list[Path],
+        ref_param="HHL",
+        ifs: bool = False,
+        delay: bool = False,
+    ):
+        """Initialize a grib reader from a list of grib files.
+
+        Parameters
+        ----------
+        datafiles : list[str]
+            List of grib input filenames
+        ref_param : str
+            name of parameter used to construct a reference grid
+        ifs : bool
+            True for setting up a grib reader for IFS data
+        delay : bool
+            if True, it will (dask) delay the functions that load parameters
+
+        Raises
+        ------
+        ValueError
+            if the grid can not be constructed from the ref_param
+
+        """
+        self._datafiles = datafiles
+        self._ifs = ifs
+        self._delayed = dask.delayed if delay else (lambda x: x)
+        if not self._ifs:
+            with cosmo_grib_defs():
```

Comment:
oh, yes, I forgot that global variable. done

Reply:
maybe this mechanism needs to be improved because it's quite easy to forget it as it is implemented now. This would be the subject for a new PR however.

Reply:
agreed